lua/plenary/filetype.lua

```diff
@@ -0,0 +1,121 @@
+
+local os_sep = Path.path.sep
+
+local filetype = {}
+
+local filetype_table = {
+  extension = {},
+  file_name = {},
+}
+
+filetype.add_table = function(new_filetypes)
+  local valid_keys = {'extension', 'file_name'}
+  local new_keys = {}
+  for k, _ in pairs(new_filetypes) do
+    new_keys[k] = true
+  end
+  for _, k in ipairs(valid_keys) do
+    new_keys[k] = nil
+  end
+
+  for k, v in pairs(new_keys) do
+    error(debug.traceback("Invalid key / value:" .. tostring(k) .. " / " .. tostring(v)))
+  end
+
+
+  if new_filetypes.extension then
+    filetype_table.extension = vim.tbl_extend("force", filetype_table.extension, new_filetypes.extension)
+  end
+
+  if new_filetypes.file_name then
+    filetype_table.file_name = vim.tbl_extend("force", filetype_table.file_name, new_filetypes.file_name)
+  end
+end
+
+filetype.add_file = function(filename)
+  local filetype_files = vim.api.nvim_get_runtime_file(
+    string.format(
+      "data/plenary/filetypes/%s.lua", filename
+    ), true
+  )
+
+  for _, file in ipairs(filetype_files) do
+    local ok, msg = pcall(filetype.add_table, dofile(file))
+    if not ok then
+      error("Unable to add file " .. file .. ":\n" .. msg)
+    end
+  end
+end
+
+filetype._get_extension = function(filename)
+  local ext_with_period = filename:match("^.+(%..+)$")
+  if ext_with_period then
+    return ext_with_period:sub(2)
+  end
+end
+
+filetype._parse_modeline = function(tail)
+  if tail:find('vim:') then
+    return tail:match('.*:ft=([^: ]*):.*$') or ''
+  end
+  return ''
+end
+
+filetype._parse_shebang = function(head)
+  if head:sub(1, 2) == '#!' then
+    return head:match('^.+/e?n?v? ?([^/ %d]*).*$') or ''
+  end
+  return ''
+end
+
+filetype.detect_from_extension = function(filepath)
+  local ext = filetype._get_extension(filepath)
+  ext = ext and ext:lower()
+  local match = ext and filetype_table.extension[ext]
+  if match then return match end
+  return ''
+end
+
+filetype.detect_from_name = function(filepath)
+  filepath = filepath:lower()
+  local split_path = vim.split(filepath, os_sep, true)
+  local fname = split_path[#split_path]
+  local match = filetype_table.file_name[fname]
+  if match then return match end
+  return ''
+end
+
+filetype.detect_from_modeline = function(filepath)
+  local tail = Path:new(filepath):tail(1)
+  return filetype._parse_modeline(tail)
+end
+
+filetype.detect_from_shebang = function(filepath)
+  local head = Path:new(filepath):head(1)
+  return filetype._parse_shebang(head)
+end
+
+filetype.detect = function(filepath)
+  local match = filetype.detect_from_name(filepath)
+  if match ~= '' then return match end
+
+  match = filetype.detect_from_extension(filepath)
+
+  if match == '' then
+    match = filetype.detect_from_shebang(filepath)
+    if match ~= '' then return match end
+  end
+
+  if match == 'text' or match == '' then
+    match = filetype.detect_from_modeline(filepath)
+    if match ~= '' then return match end
+  end
+
+  return match
+end
```

Comment:
I split the whole thing.
So `detect` will do all `name -> extension -> shebang (if extension '') -> modeline (if extension text or '')`.
And then i have a interface for `extension`, `name`, `shebang`, `modeline` if some people want only one or two of them.